### ISSUE TITLE
map_to with extra context

### ISSUE THREAD

**Issue by filwaline** [dataclass-mapper/dataclass-mapper#8]
`mapping` parameter of `mapper` and `mapper_from` can not provide data other from source-object, so missing field of target-class need default value and modify after mapping.

map and modify, this pattern work, but not the best. if modification not validate properly, it sure breaks something. 

it is easier to create a [Always-Valid Domain Model](https://enterprisecraftsmanship.com/posts/always-valid-domain-model/) at first, rather than validate modification later.

why not enable `map_to` carry extra data, so target-class can validate everything at beginning?

```python
from pydantic import BaseModel
import dataclass_mapper


class ComplexObject(BaseModel):
    ...


class SourceModel(BaseModel):
    id: str
    data1: str


@dataclass_mapper.mapper_from(
    SourceModel,
    {
        "data2": "dataclass_mapper.provide_at_mapping()"
    }
)
class TargetModel(BaseModel):
    id: str
    data1: str
    data2: ComplexObject


source = SourceModel(id="id", data1="any data")
target = dataclass_mapper.map_to(
    source, TargetModel, extra={'data2': ...}
)
```

**Comment by filwaline** [comment]
```python
from pydantic import BaseModel
import dataclass_mapper


class ComplexObject(BaseModel):
    ...


class SourceModel(BaseModel):
    id: str
    data1: str

@dataclass_mapper.mapper_from(SourceModel,{"data2": dataclass_mapper.init_with_default()})
class TargetModel2(BaseModel):
    id: str
    data1: str
    data2: ComplexObject = Field(default_factory=...)

    class Config:
        validate_assignment = True

target2 = dataclass_mapper.map_to(source, TargetModel2)
target2.data2 = ...  # not so good
```
this works, but i don't like it.

**Comment by jakobkogler** [comment]
Good idea.
When I started writing this library, I had something similar in my mind. But wanted to focus on other aspects first, and didn't had any good ideas yet about the syntax.

Your suggestion sounds nice. Although I'm not yet sure how it will work with recursive models.
One idea would be the following. What do you like it?
```py
class TargetItem(BaseModel):
    x: int
    y: int

class Target(BaseModel):
    item: TargetItem
    items: list[TargetItem]
    optional_item: Optional[TargetItem]

@mapper(TargetItem, {"y": provide_at_mapping()})
class SourceItem(BaseModel):
    x: int

@mapper(Target):
class Source(BaseModel):
    item: SourceItem
    items: list[SourceItem]
    optional_item: Optional[SourceItem]

source = Source(
    item=SourceItem(x=1), 
    items=[SourceItem(x=2), SourceItem(x=3)]
    optional_item=None
)
target = map_to(
    source, 
    Target, 
    extra={
        "item": {"y": 9},
        "items": [
            {"y": 8},
            {"y": 7}
        ]
        # and additionally `"optional_item": {"y": 6}` if `source.optional_item` is not None
    }
)
```

**Comment by jakobkogler** [comment]
Implemented in 365299e177ecaf51146c9af807250344407f9f55
Released with version 1.6.0
